File: bug18_wrong_index_position/solved_main.py
import base64, struct

def run():
    tbl = [i * i * 0x1337 for i in range(1, 11)]
    acc = 0xC0FFEEDEADBEEF42
    for step in range(10):
        v = tbl[9 - step]  # FIX: was "9 - step + 1" which causes index 10 (out of range); remove the +1
        acc = (acc ^ v) * 0x9E3779B97F4A7C15
        acc &= 0xFFFFFFFFFFFFFFFF
        acc = ((acc << (step % 16 + 1)) | (acc >> (63 - step % 16))) & 0xFFFFFFFFFFFFFFFF
    raw = struct.pack(">Q", acc)
    print(base64.b64encode(raw).decode())

run()

File: bug18_wrong_index_position/solved_main.js
function run() {
    let tbl = [];
    for (let i=1n; i<=10n; i++) tbl.push(i*i*0x1337n);
    let acc = 0xC0FFEEDEADBEEF42n;
    for (let step=0; step<10; step++) {
        let idx = 9 - step;  // FIX: was (9-step+1) which is out of bounds at step=0; remove +1
        let v = tbl[idx];
        acc = BigInt.asUintN(64, (acc^v)*0x9E3779B97F4A7C15n);
        let sl=BigInt(step%16+1), sr=BigInt(63-step%16);
        acc = BigInt.asUintN(64, (acc<<sl)|(acc>>sr));
    }
    let buf=Buffer.alloc(8); buf.writeBigUInt64BE(acc,0);
    console.log(buf.toString('base64'));
}
run();
